src/SocioCatClaim.sol

```diff
@@ -3,20 +3,51 @@ pragma solidity =0.8.25;
 
 import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {SignatureChecker} from "@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol";
 
 contract SocioCatClaim {
     using SafeERC20 for IERC20;
+
     IERC20 public immutable token;
+    address public signer;
+    mapping(uint256 => bool) public nonceUsed;
+
+    error InvalidSignature();
+    error InvalidNonce();
 
-    constructor(IERC20 _token) {
+    event Claimed(address indexed to, uint256 amount);
+
+    constructor(IERC20 _token, address _signer) {
+        signer = _signer;
         token = _token;
     }
 
-    function claim(uint256 amount, address receiver) external {
+    function claim(
+        uint256 amount,
+        uint256 nonce,
+        bytes calldata signature,
+        address receiver
+    ) external {
+        if (nonceUsed[nonce]) {
+            revert InvalidNonce();
+        }
```

Comment:
We might end up signing the same claim twice with different nonces, which means users could snag more CAT tokens than they're supposed to.

So, how about we sign this stuff instead:

- recipient
- maxClaimAmount
- amount

And then, we keep track of a `recipient -> claimedAmount` map.

Users can make a claim if:
- Signature is valid
- `claimedAmount + amount <= maxAmount`

This method should let us sign off on things without stressing about nonces

Reply:
make sense 🚀